rust/datafusion/src/logical_plan/mod.rs

```diff
@@ -323,21 +322,19 @@ impl Expr {
     ///
     /// # Errors
     ///
-    /// This function errors when it is impossible to cast the expression to the target [arrow::datatypes::DataType].
+    /// Currently no errors happen at plan time. If it is impossible
```

Comment:
It is interesting that there appear to be no uses (or tests) for this function -- https://github.com/apache/arrow/search?q=cast_to I found this quite confusing

@jorgecarleitao  what do you think about simply removing the whole thing entirely?

Reply:
I am sorry, I was confused: `can_coerce_from` checks whether we can perform a lossless cast. This is different from the ability to perform *any* cast.

Therefore, I think that the logic we may want is:

* the `Expr::Cast` operation should check that we can perform the cast at all (e.g. a `ListArray` to a `UInt32Array` should not be allowed). This should match the casts that the kernel supports.
* implicit casts, such as the ones that we perform when we pass arguments to functions to try to match the signature, should use `can_coerce_from`, as we do not want to perform lossless implicit casts.

AFAIK casts to and from dictionaries are lossless, and thus they can happen at any point in the physical planning. Since all arrays support dictionaries, I guess we only have to error if the implementation is still not available.

What do you think, @alamb ?

Reply:
@jorgecarleitao -- I think this makes sense. I was confused about the intent of `can_coerce_from` (aka that it is for lossless conversions only). With this explanation let me go back and give adding dictionary coercion / cast a second pass (though likely not until tomorrow, Thursday, US Eastern time)

Reply:
@jorgecarleitao  -- here is a proposed alternate approach: https://github.com/apache/arrow/pull/8400